docs/closure-compiler.md

```diff
@@ -0,0 +1,699 @@
+# Annotating MDC-Web for the Closure Compiler
+
+> TL;DR read the section on our [type system](#mdc-web-type-system) and our [conventions](#mdc-web-closure-conventions).
+
+## Who this document is for
+
+This document is for _core contributors to MDC-Web, as well as contributors who wish to author
+new components, or make non-trivial changes to existing components._ It assumes you're familiar with
+our codebase, and have read through most of our [Authoring Components guide](./authoring-components.md).
+
+## Why this is needed
+
+MDC-Web - and Material Design in general - was created by Google. Therefore, it is not only a top
+priority that MDC-Web works seamlessly for our external community, but also that _MDC-Web works
+seamlessly for all Google applications_. This is an absolute requirement for us.
+
+At Google, all Javascript is processed and minified by the
+[Closure Compiler](https://github.com/google/closure-compiler) (which will be referred to as
+**closure**, the **compiler**, or any combination of those terms). Thus, _in order for every Google
+application to deem MDC-Web viable for use within it, the library must be compilable using
+closure's [advanced compilation](https://developers.google.com/closure/compiler/docs/api-tutorial3)
+mechanisms_.
+
+### What about externs?
```

Comment:
Definitely, especially from Googlers. We've been asked about this internally a few times which is why I included it. I can make "externs" a link so that anyone unfamiliar can find out more info.